frontend/src/metabase/static-viz/components/LineAreaBarChart/utils/series.tsx

```diff
@@ -0,0 +1,174 @@
+import _ from "underscore";
+import { merge } from "icepick";
+import { isNotNull } from "metabase/core/utils/types";
+import { getColorsForValues } from "metabase/lib/colors/charts";
+import { formatStaticValue } from "metabase/static-viz/lib/format";
+import { ColorPalette } from "metabase/lib/colors/types";
+import {
+  ChartSettings,
+  Series,
+  SeriesWithOneOrLessDimensions,
+  SeriesWithTwoDimensions,
+} from "../../XYChart/types";
+
+export function getSeriesWithColors(
+  multipleSeries: (SeriesWithOneOrLessDimensions | SeriesWithTwoDimensions)[][],
+  settings: ChartSettings,
+  palette: ColorPalette,
+): (SeriesWithOneOrLessDimensions | SeriesWithTwoDimensions)[][] {
+  const isMultipleSeries = multipleSeries.length > 1;
+  const keys = getSeriesKeys(multipleSeries, isMultipleSeries);
+
+  const seriesColors = settings.series_settings
+    ? _.mapObject(settings.series_settings, value => {
+        return value.color;
+      })
+    : undefined;
+  const chartColors = getColorsForValues(
+    keys,
+    removeEmptyValues(seriesColors),
+    palette,
+  );
+
+  let index = -1;
+  return multipleSeries.map(questionSeries =>
+    questionSeries.map(series => {
+      index++;
+
+      return merge(series, {
+        color: chartColors[keys[index]],
+      });
+    }),
+  );
+}
+
+export function getSeriesWithLegends(
+  multipleSeries: (SeriesWithOneOrLessDimensions | SeriesWithTwoDimensions)[][],
+  settings: ChartSettings,
+): (SeriesWithOneOrLessDimensions | SeriesWithTwoDimensions)[][] {
+  const keys = getSeriesKeys(multipleSeries, multipleSeries.length > 1);
+  const isMultipleSeries = multipleSeries.length > 1;
+
+  const seriesTitles = settings.series_settings
+    ? _.mapObject(settings.series_settings, value => {
+        return value.title;
+      })
+    : undefined;
+
+  let index = -1;
+  const legends = multipleSeries
+    .flatMap((questionSeries, seriesIndex) => {
+      return questionSeries.map(series => {
+        index++;
+
+        if (!hasTwoDimensions(series)) {
+          // One or zero dimensions
+
+          if (seriesIndex === 0 && series.name) {
+            return series.name;
+          }
+
+          const hasOneMetric = questionSeries.length === 1;
+          if (hasOneMetric) {
+            return series.cardName;
+          }
+
+          if (!isMultipleSeries) {
+            return series.column.display_name;
+          }
+
+          // is multiple series card
+          return `${series.cardName}: ${series.column.display_name}`;
+        } else {
+          // Two dimensions
+
+          const columnKey = formatStaticValue(series.breakoutValue, {
```

Comment:
More explanation for this line here
[placeholder for a comment in `src/metabase/pulse/render/body.clj`]